invoices/invoiceregistry.go

```diff
@@ -1260,11 +1260,30 @@ func (i *InvoiceRegistry) SubscribeNotifications(
 		}
 	}()
 
+	// Query the database to see if based on the provided addIndex and
+	// settledIndex we need to deliver any backlog notifications.
+	i.Lock()
+	addEvents, err := i.cdb.InvoicesAddedSince(addIndex)
+	if err != nil {
+		i.Unlock()
+		return nil, err
+	}
+	client.addBacklog = addEvents
+
+	settleEvents, err := i.cdb.InvoicesSettledSince(settleIndex)
+	if err != nil {
+		i.Unlock()
+		return nil, err
+	}
+	client.settleBacklog = settleEvents
+
 	select {
 	case i.newSubscriptions <- client:
 	case <-i.quit:
+		i.Unlock()
```

Comment:
Doesnt this have deadlock potential? Waiting for the subscription to be picked up by the main loop inside the lock?

Reply:
in theory yes, but as long  as we are consistent it is fine. this is already what happens with `AddInvoice`, `SettleInvoice`, `CancelInvoice`, `notifyClients` is called with the mutex held, which feeds an event via `invoiceEvents` into the main loop.

Reply:
In the latest version of this diff, the main loop doesn't need to grab any locks when a new subscription is added.